numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -2019,6 +2020,60 @@ static PyObject *
     Py_INCREF(ret);
     return ret;
 }
+
+static PyObject *
+@name@_hex(PyObject *self) {
+#if @is_half@
+    npy_double val = npy_half_to_double(PyArrayScalar_VAL(self, @Name@));
+#else
+    npy_@name@ val = PyArrayScalar_VAL(self, @Name@);
+#endif
+    // xref: https://stackoverflow.com/a/15197157/5671364
+    char buf[
+        1 + // sign, '-' or '+'
+        (NPY_SIZEOF_@NAME@ * CHAR_BIT + 3) / 4 + // mantissa hex digits max
+        1 + // decimal point, '.'
+        1 + // mantissa-exponent separator, 'p'
+        1 + // mantissa sign, '-' or '+'
+        (NPY_SIZEOF_@NAME@ * CHAR_BIT + 2) / 3 + // exponent decimal digits max
+        1 // string terminator, '\0'
+    ];
+
+    if (npy_isnan(val) || !npy_isfinite(val)) {
+        return floattype_str(self);
+    }
+    npy_@name@_hex(val, buf);
+    return PyUnicode_FromString(buf);
+}
+
+static PyObject *
+@name@_fromhex(PyObject *cls, PyObject *arg) {
+    const char *string;
+    Py_ssize_t length;
+    PyObject *ret;
+#if @is_half@
+    npy_double value;
+#else
+    npy_@name@ value;
+#endif
+
+    string = PyUnicode_AsUTF8AndSize(arg, &length);
+    if (string == NULL) {
+        return NULL;
+    }
+
+    if (npy_@name@_fromhex(string, &value) == NPY_FALSE) {
+        return NULL;
+    }
+
+    ret = PyArrayScalar_New(@Name@);
+    if (ret == NULL) {
+        return NULL;
+    }
+    PyArrayScalar_ASSIGN(ret, @Name@, value);
```

Comment:
Hey @seberg / @mattip / @eric-wieser , apologies for the ping, I could not figure this out. This logic is not working for `flaot16` for some reason. For input say `>>> np.float16.fromhex(np.float16(2.0).hex())`, this is the trace:
```c
2073        PyArrayScalar_ASSIGN(ret, @Name@, value);
(gdb) 
2075        return ret;
(gdb) p value
$1 = 2
(gdb) c
Continuing.
1e-07
```
It seems like result is `1e-07` although the logic gives result(`value`) as `2`. Any pointers here will be very helpful. Thanks.

Reply:
The half path uses double (a bit weird that not float?), and you need to convert float/double to half first to store it, since there is no C `half` type.

Reply:
Ohh I see. Shall I use float? I had the same doubt but referred to here where double was used and wondered why: https://github.com/numpy/numpy/blob/c8fdb53ee2d1b1130a74ad7a3250366fd042da3f/numpy/core/src/multiarray/scalartypes.c.src#L2005
Thanks a lot for the reply!

Reply:
Doesn't matter probably, if you are pretty sure there is no issue (and why would there be?) changing to float seems reasonable, since that is what most places do AFAIK.

Reply:
It worked, thanks a lot!